### PR TITLE
add sourmash to private and public CLI

### PR DESCRIPTION
**Pull request by kiepczi**
This follows from #166, where we added the initial `sourmash` workflow.

In this PR, I:
- wrote an initial parsing method for sourmash compare to extract estimated ANI values,
- provided logging to the DB via the private CLI, and
- added `sourmash` to the public CLI.

The logic for logging to the DB via the private CLI and adding `sourmash` to the public CLI follows the approach used in other methods like `aniM`, `fastANI`  etc.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [X] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)
- [ ] This change requires a documentation update
- [ ] This is a documentation update

## Action Checklist

- [X] Work on a single issue/concept (if there are multiple separate issues to address, please use a separate pull request for each)
- [X] Fork the `pyani-plus` repository under your own account (please [allow write access for repository maintainers](https://docs.github.com/en/github/collaborating-with-issues-and-pull-requests/working-with-forks/allowing-changes-to-a-pull-request-branch-created-from-a-fork))
- [X] Set up an appropriate development environment (please see `CONTRIBUTING.md` for this repository)
- [X] Create a new branch with a short, descriptive name
- [X] Work on this branch
  - [X] style guidelines have been followed (please see `CONTRIBUTING.md` for this repository)
  - [X] new code is commented, especially in hard-to-understand areas
  - [ ] corresponding changes to documentation have been made
  - [X] tests for the change have been added that demonstrate the fix or feature works
- [X] Test locally with `pytest-plus -v` **non-passing code will not be merged**
- [ ] Rebase against `origin/master`
- [X] Check changes with linting/formatting tools before submission (please see `CONTRIBUTING.md` for this repository)
- [X] Commit branch
- [X] Submit pull request, describing the content and intent of the pull request
- [X] Request a code review
- [X] Continue the discussion at [`Pull requests` section](https://github.com/widdowquinn/pyani-plus/pulls) in the `pyan-plus` repository

### CODE REVIEW

**Comment by kiepczi** [comment]
Things that we still need to discuss/change:
- Currently, with sourmash, we are only able to report estimated ANI values, not coverage, the number of aligned bases, or similarity errors. I need to investigate if there is a way to retrieve this information.
- In the public CLI, the argument extra is set to scaled=300 for now, but this will need to be changed to scaled=1000 to match the default settings in sourmash. It is currently set to scaled=300 because scaled=1000 fails to estimate ANI values for our set of viral genomes. I can't figure out how to change the parse settings to scaled=300 for testing when the default is set to scaled=1000.

**Comment by peterjc** [comment]
It appears there is no built-in way in ``typer`` to do mutually exclusive options (although this is possible in ``argparse``), although there are various workarounds - none actually show this in the help. I would therefore keep in simple and define two optional integer arguments and check within the function body:

```python
@app.command(rich_help_panel="ANI methods")
def sourmash(  # noqa: PLR0913
    fasta: REQ_ARG_TYPE_FASTA_DIR,
    database: REQ_ARG_TYPE_DATABASE,
    # These are for the run table:
    name: REQ_ARG_TYPE_RUN_NAME,
    *,
    # The mode here is not optional - must pick one!
    mode: OPT_ARG_TYPE_SOURMASH_MODE = method_sourmash.MODE,
    create_db: OPT_ARG_TYPE_CREATE_DB = False,
    executor: OPT_ARG_TYPE_EXECUTOR = ToolExecutor.local,
    scaled: OPT_ARG_TYPE_SOURMASH_SCALED = method_sourmash.SCALED,  # this is 1000
    num: OPT_ARG_TYPE_SOURMASH_NUM = None,
    kmersize: OPT_ARG_TYPE_KMERSIZE = method_sourmash.KMER_SIZE,
) -> int:
    """Execute sourmash calculations, logged to a pyANI-plus SQLite3 database."""
    if scaled and num:
        msg = "ERROR: Options --scaled and --num are mutually exclusive."
        sys.exit(msg)
    elif scaled:
        extra = f"scaled={extra}"
    else:
        extra = f"num={num}"
    check_db(database, create_db)
    ...
```
```

File: tests/test_sourmash.py [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+        compare=sourmash_targets_compare_indir
+        / "MGV-GENOME-0264574_vs_MGV-GENOME-0266457.csv",
+    )
+
+    # Check the recorded comparison values
+    session = db_orm.connect_to_db(tmp_db)
+    assert session.query(db_orm.Comparison).count() == 1
+    comp = session.query(db_orm.Comparison).one()
+    query = "689d3fd6881db36b5e08329cf23cecdd"  # MGV-GENOME-0264574.fas
+    subject = "78975d5144a1cd12e98898d573cf6536"  # MGV-GENOME-0266457.fna
```
Comment (peterjc) [inline tests/test_sourmash.py:150]: You can confirm ``query == comp.query_hash`` here (and for subject), or remove those two lines and use ``comp.query_hash`` below?

File: tests/test_public_cli.py [hunk 2]
```diff
@@ -337,6 +337,22 @@ def test_fastani(
...
 
+def test_sourmash(
+    tmp_path: str, input_genomes_tiny: Path, dir_sourmash_results: Path
+) -> None:
+    """Check sourmash run (default settings)."""
+    out = Path(tmp_path)
+    tmp_db = out / "example.sqlite"
+    public_cli.sourmash(
+        database=tmp_db, fasta=input_genomes_tiny, name="Test Run", create_db=True
+    )
```
Comment (peterjc) [inline tests/test_public_cli.py:348]: Here you should be overriding the defaults for scaled/num (which won't work for the viral genomes) with the values used to generate the fixtures.

**Comment by peterjc** [comment]
Mutually exclusive arguments and default values is fiddly. Comment above updated with two ways that would work, with slightly different end-user experiences.

File: tests/generate_fixtures/generate_target_sourmash_matrices.py [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+# THE SOFTWARE.
+"""Generate target matrices for pyani-plus sourmash tests.
+
+This script can be run with ``./generate_target_anim_files.py`` in the
```
Comment (peterjc) [inline tests/generate_fixtures/generate_target_sourmash_matrices.py:26]: Name needs updating (anim to sourmash)

File: tests/generate_fixtures/generate_target_sourmash_matrices.py [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+script's directory, or from the project root directory via ``make fixtures``.
+It will regenerate and potentially modify test input files under the
+fixtures directory.
+
+This script generates target matrices for dnadiff method comparisons from
+sourmash compare .csv files.
```
Comment (peterjc) [inline tests/generate_fixtures/generate_target_sourmash_matrices.py:32]: Again a stale reference but to dnadiff

File: tests/generate_fixtures/generate_target_sourmash_matrices.py (hunk 3 above)
Comment (peterjc) [reply tests/generate_fixtures/generate_target_sourmash_matrices.py:26]: Also this needs to be added to the ``Makefile``

**Comment by kiepczi** [comment]
Thanks so much, @peterjc, for the such detailed review—really appreciate it, especially the tips on handling two mutually exclusive arguments. It would’ve taken me ages to figure that out on my own, and it’s great to see how proper (and elegant!) argument parsing is done.

I'll get to this as soon as I can, as I still need to prep for tomorrow's talk.

**Comment by kiepczi** [comment]
> Also we need to add `sourmash` to the third-party requirements - any particular version?


Oh sugar! I added `sourmash` to `requirements.txt` rather than the third-party requirements. I will change this now. I have been working on the newest `sourmash` version (4.8.11), so I guess I will set this as the minimum for now?

**Comment by peterjc** [comment]
We should put a comment block at the top of ``requirements.txt`` that is it python-only things (and in principle should work with pip install or conda install), like https://github.com/peterjc/thapbi-pict/blob/master/requirements.txt

**Comment by kiepczi** [comment]
Hi @peterjc. Unfortunately, I wasn't able to finish all of this before going on annual leave. I lost a bit of time trying to get the MacStudio to behave again.

I will be back in the office on the 18th, so don't worry if you can't get to this by then. I can handle what’s still left to do when I'm back.

**Comment by peterjc** [comment]
That's great - have a good break 👋

**Comment by peterjc** [comment]
See 35d335ce45ce9f1f5ac72e54be4a8a76f702ba45 for updating the repr.

**Pull request merged**